010-develop/065-quantization.md

```diff
@@ -2,8 +2,255 @@
+
+Product quantization can achieve greater memory savings than binary and scalar quantization, but less computation savings.
+
+### Basic Usage
+
+The first step of product quantization is the `quantize_table` function. This function creates a codebook for each subvector and quantizes the vectors in the table into a column with the same name as the original column with `_pq` appended.
+
+For example, the following command quantizes the table `books` with `v` as the vector column, and generates a column `v_pq` with 256 clusters, 16 subvectors, and the Euclidean distance metric:
+
+---
+
+```sql
+SELECT quantize_table('books', 'v', 256, 16, 'l2sq');
+```
+
+To create the index, set `pq=True` in the `WITH` clause of the index creation statement:
+
+```sql
+CREATE INDEX hnsw_l2_index ON small_world_pq USING lantern_hnsw(v) WITH (pq=True);
+```
+
+Finally, to query the index, you can use the usual query operator. There is no need to convert the query vector to the product-quantized format, as this is handled internally by Lantern.
+
+```sql
+SELECT * FROM books ORDER BY v <-> '[1,2,3]' LIMIT 5;
+```
+
+The full list of available functions for product quantization follows.
+
+## Product Quantization API
+
+The following functions are available in SQL for product quantization:
+
+### quantize_table
+
+This function creates a codebook for each subvector and quantizes the vectors in the table into a column with the same name as the original column with `_pq` appended.
+
+The API details are as follows:
+
+```sql
+quantize_table(p_tbl regclass, p_col NAME, cluster_cnt INT, subvector_count INT, distance_metric TEXT, dataset_size_limit INT DEFAULT 0);
+```
+
+---
+
+```documentation
+p_tbl
+regclass
+```
+
+The table to quantize.
+
+---
+
+```documentation
+p_col
+NAME
+```
+
+The column to quantize.
+
+---
+
+```documentation
+cluster_cnt
+INT
+```
+
+The number of clusters to use for each subvector.
+
+---
+
+```documentation
+subvector_count
+INT
+```
+
+The number of subvectors to split the vector into.
+
+---
+
+```documentation
+distance_metric
+TEXT
+```
+
+The distance metric to use for quantization. The supported distance metrics are `l2sq`, `cos`, and `hamming`.
+
+---
+
+```documentation
+dataset_size_limit
+INT
+```
+
+k-means clustering is a memory-intensive operation. If the dataset is too large, the clustering operation may fail due to memory constraints. To avoid this, you can set a limit on the number of rows to use for clustering. If the dataset size exceeds this limit, the function will sample the dataset to the specified limit.
+
+### create_pq_codebook
+
+Unlike `quantize_table`, `create_pq_codebook` only creates the codebook without creating the column It has the same arguments as `quantize_table`.
+
+### quantize_vector
+
+Unlike `quantize_table`, `quantize_vector` creates a new column with the quantized vectors.
```

Comment:
Actually this function doesn't create a column but returns quantized value for the given vector